File: bindings/java/src/main/java/systems/zlink/contracts/sockets/SocketEnums/CompletionKind.java
/* SPDX-License-Identifier: MPL-2.0 */

package systems.zlink.contracts.sockets;

/** Identifies the kind of a native socket completion record. */
public enum CompletionKind {
    /**
     * ABI-retained legacy SEND value. Ordinary successful sends do not emit a
     * SEND completion.
     */
    SEND(1),
    /** A request reached its asynchronous terminal result. */
    REQUEST(2),
    /**
     * Resolves a previously backpressured send token. A writable result allows
     * the same packet to be retried; terminal fields report shutdown instead.
     */
    WRITABLE(3);

    private final int value;

    CompletionKind(int value) {
        this.value = value;
    }

    public int value() {
        return value;
    }

    public static CompletionKind fromValue(int value) {
        return switch (value) {
            case 1 -> SEND;
            case 2 -> REQUEST;
            case 3 -> WRITABLE;
            default -> throw new IllegalArgumentException(
                "invalid CompletionKind value: " + value);
        };
    }
}

File: bindings/dotnet/src/Zlink/Contracts/Messaging/CompletionKind.cs
// SPDX-License-Identifier: MPL-2.0

namespace Systems.Zlink;

/// <summary>Identifies records stored in a socket completion queue.</summary>
public enum CompletionKind
{
    /// <summary>
    ///     ABI-reserved SEND completion value. Successful SEND admission uses
    ///     completion ID 0 and does not enqueue a completion record.
    /// </summary>
    Send = 1,

    /// <summary>A request completed with a reply or terminal result.</summary>
    Request = 2,

    /// <summary>
    ///     A DONTWAIT SEND wait token became writable. This grants permission to
    ///     retry the same packet; it does not report SEND success.
    /// </summary>
    Writable = 3
}
